File: 240804/겹치지 않게 선분 고르기/select-segments-without-overlap.py
n = int(input())
lines = []

for _ in range(n):
    a, b = tuple(map(int, input().split()))
    lines.append((a, b))

ans = []

def max_lines():

    paper = dict()

    for i in range(n):
        if ans[i] == 1:
            start, end = lines[i]
        else:
            continue
        
        for num in range(start, end+1):
            if num not in paper:
                paper[num] = 1
                # print(paper)
            else:
                # print(paper)
                return 0
    result = 0
    

    for a in ans:
        if a == 1:
            result += 1
        
    return result 





def print_ans():
    for a in ans:
        print(a, end="")

    print()

result = 0
def choose(num):
    global result 

    if num == n:
        # print_ans()
        result = max(result, max_lines())
        # print(result)
        return 

    for i in range(2):
        ans.append(i)
        choose(num+1)
        ans.pop()

choose(0)
print(result)

File: 241004/겹치지 않게 선분 고르기/select-segments-without-overlap.cpp
#include <iostream>
#include <tuple>
#include <vector>

using namespace std;

int n; 
vector<pair<int, int>> selected; 
vector<pair<int, int>> lines; 


// void print_ans(){
//     for(auto a : ans)
//         cout << a << " ";
//     cout << endl; 
// }



bool overlapped(pair<int, int> l1, pair<int,int> l2){
    int a1, a2;
    tie(a1, a2) = l1;

    int b1, b2;
    tie(b1, b2) = l2;

    return  (a1 <= b1 && b1 <= a2) || (a1 <= b2 && b2 <= a2) ||
            (b1 <= a1 && a1 <= b2) || (b1 <= a2 && a2 <= b2) ;
}

bool possible(){
    for(int i = 0; i<(int)selected.size(); i++){
        for(int j = i+1; j<(int)selected.size(); j++){
            if (overlapped(selected[i], selected[j]))
                return false;
        }
    }
    return true;
}


void print_selected(){
    for(auto s : selected)
        cout << s.first << " " << s.second << " " << "|" <<" ";
    cout << endl;
}

int result = 0; 

void choose(int cnt){
    if (cnt == n){  
        if (possible()){
            // print_selected();
            result = max(result, (int)selected.size());
        }

        
        return;
    }

    selected.push_back(lines[cnt]);
    choose(cnt+1);
    selected.pop_back();

    choose(cnt+1);

}

int main() {

    cin >> n; 
    for (int i = 0; i < n; i++){
        int x1, x2;
        cin >> x1 >> x2; 
        lines.push_back(make_pair(x1, x2)); 
    }

    choose(0);
    cout << result;
    return 0;
}